app/Http/Controllers/App/UsersController.php

```diff
@@ -3,30 +3,42 @@
 namespace Adshares\Adserver\Http\Controllers\App;
 
 use Adshares\Adserver\Mail\UserEmailActivate;
+use Adshares\Adserver\Mail\UserEmailChangeConfirm1Old;
+use Adshares\Adserver\Mail\UserEmailChangeConfirm2New;
 use Adshares\Adserver\Models\User;
+use Adshares\Adserver\Models\Token;
 use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\DB;
 use Illuminate\Support\Facades\Mail;
+use Illuminate\Support\Facades\Validator;
 
 class UsersController extends AppController
 {
+    protected $email_activation_token_time = 24 * 60 * 60; // 24 hours
+    protected $email_activation_resend_limit = 15 * 60; // 15 minutes
+    protected $email_change_token_time = 60 * 60; // 1 hour
+
     /**
      * Create a new controller instance.
      */
     public function __construct()
     {
-        $this->middleware('snake_casing');
+        $this->middleware('snake_casing')->except(['emailChangeStep1']);
     }
 
     public function add(Request $request)
     {
-        $this->validateRequest($request, 'user', User::$rules_add);
-        $user = new User($request->input('user'));
-        $user->password = $request->input('user.password');
-        $user->email = $request->input('user.email');
-        $user->email_confirm_token = md5(openssl_random_pseudo_bytes(20));
-        $user->save();
+        $this->validateRequestObject($request, 'user', User::$rules_add);
+        Validator::make($request->all(), ['uri' => 'required'])->validate();
 
-        Mail::to($user)->queue(new UserEmailActivate($user));
+        DB::beginTransaction();
+        $user = User::register($request->input('user'));
+        Mail::to($user)->queue(new UserEmailActivate(
+            Token::generate('email-activate', $this->email_activation_token_time, $user->id),
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "email-activate" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)